### DISCUSSION TITLE
How do I write a language pack for Theia?

### DISCUSSION THREAD

**Discussion by oleks** [eclipse-theia/theia#13507] [Q&A]
The [Theia internationalization documentation](https://theia-ide.org/docs/i18n) mentions that one can use [Visual Studio Code language packs](https://marketplace.visualstudio.com/search?target=VSCode&category=Language%20Packs&sortBy=Installs) to localize my Theia. However, that is a **severely limited** list of languages.

What is the best way to add a language pack for a language not listed there?

**Comment by msujew** [comment]
Hey @oleks,

You can find all language packs of vscode in [this repo](https://github.com/microsoft/vscode-loc). They are basically just large json files that provide translations for all values.

**Reply by oleks** [reply]
It is not enough to just write a JSON file, as evidenced by [this discussion](https://github.com/microsoft/vscode-loc/issues/1424#issuecomment-1634675274). I have previously been involved in [trying and giving up](https://github.com/microsoft/vscode-loc/pull/1239) on adding a new language pack to the vscode collection. I was hoping that using Theia would be easier.

**Reply by oleks** [reply]
To be more concrete, they don't take the JSON files as a source for the translations. Instead, the JSON files simply mirror the data of an "internal tool".

As evidenced by the comment `Do not edit this file. It is machine generated.` at the top of these JSON files.

**Reply by msujew** [reply]
Hm, yeah I can see that. I implemented the i18n capabilities for Theia back in '21, but the only use cases we had in mind was to support the existing language packs of vscode and adding additional translations that are custom to Theia extensions.

With that in mind, I recommended to look at the json files, since we don't have a tool to generate something similar ourselves. We have a tool for building translations for Theia extensions, but nothing for vscode language packs. So aside from that, I can't really recommend anything.
